### ISSUE TITLE
Auxiliary cost - reconstruction part

### ISSUE THREAD

**Issue by hksBRT** [sordonia/zforcing#1]
For the reconstruction part eq 8, it is shown that b_t is obtained from z_t, but in the code you have concatenated h_t and z_t to get the mean and logvar for b_t. Was there a reason for doing this?

**Comment by sordonia** [comment]
Hi, sorry for the late answer , I was away in a conference . In principle the concatenation of h_t and z_t allows to have a more powerful density model of backward states. We experimented with both, but we didn't find significant differences in the final performance. In the code the z_force option sets h_t to 0. effectively giving Eq. 8 in the paper

**Comment by hksBRT** [comment]
Thanks for the clarification. I have two more doubts in the code:
1. What is the reason for using a custom LSTM cell? I believe the cond_ln refers to layer norm. If I am interested in implementing the paper in tensorflow referring to your code, can I use a tf.contrib.rnn.LayerNormBasicLSTMCell?
2. For the torch clamp operation, is there a reason for choosing +-8 as the limits?

**Comment by sordonia** [comment]
1. I am not very familiar with tensorflow but i think so

2. Not a particular reason, but I found them to be beneficial to stabilize training

**Comment by hksBRT** [comment]
I am running to run your code in pytorch and it fails at the rnnFusedPointwise in the backend part. The type2backend.load gives me a KeyError <class torch.Tensor'>

**Comment by sordonia** [comment]
yeah, no idea. can you try with pytorch 0.4.1 ?

**Comment by hksBRT** [comment]
Yup, that works. Thanks

**Comment by hksBRT** [comment]
Where can I access the Timit or Blizzard dataset used for this paper so that I can reproduce the results from the paper?

**Comment by sordonia** [comment]
Data cannot be released :-( it needs to be obtained from https://catalog.ldc.upenn.edu/LDC93S1

**Comment by hksBRT** [comment]
On running your model on the TIMIT dataset, i think I get similar results for the KL divergence plot as in Fig2. My final kld value is 2451 for the model + KLA. However, my final nll for the test is -66505. Infact my nll starts from +1000 in epoch 0 and goes to this high -ve value at the end of training. It looks like the paper reports the log likelihood for the test data, so the negative of my final nll closely reflects the value mentioned in the paper. 
Would really appreciate if you could answer the following questions:
1. The final test loss is a sum of the fwd nll and kld. Is that the same even when I use aux weight annealing?
2. Could you also point to how the aux weight annealing is done? Is it similar to the kld annealing (0.2 to 1 with step of 0.00005)? 
3. Is the ideal hyper-parameter setting the default one?
4. For train and evaluate, shouldn't the init_hidden be called inside the inner most for loop (get_batch)
5. For my experiments with another dataset, I see that the KLD approaches to 0 and starts diverging. Theoretically, I expected that to happen since the prior and posterior start coming closer. Why does the KLD increase as explained in your paper?

**Comment by hksBRT** [comment]
Also, why do you mask the test dataset?
